aws/resource_aws_kms_grant.go

```diff
@@ -0,0 +1,479 @@
+// Use a marker to iterate over all grants in "pages"
+// NB: This function only retries on AWS Errors
+func findKmsGrantById(conn *kms.KMS, keyId string, grantId string, marker *string) (*kms.GrantListEntry, error) {
+
+	input := kms.ListGrantsInput{
+		KeyId:  aws.String(keyId),
+		Limit:  aws.Int64(int64(100)),
+		Marker: marker,
+	}
+
+	var out *kms.ListGrantsResponse
+	var err error
+	var grant *kms.GrantListEntry
+
+	err = resource.Retry(3*time.Minute, func() *resource.RetryError {
+		out, err = conn.ListGrants(&input)
+
+		if err != nil {
+			if awsErr, ok := err.(awserr.Error); ok {
+				if awsErr.Code() == "NotFoundException" ||
+					awsErr.Code() == "DependencyTimeoutException" ||
+					awsErr.Code() == "InternalException" {
+					return resource.RetryableError(err)
+				}
+			}
+			return resource.NonRetryableError(err)
+		}
+
+		return nil
+	})
+
+	grant = getKmsGrantById(out.Grants, grantId)
+	if grant != nil {
+		return grant, nil
+	}
+	if *out.Truncated {
+		log.Printf("[DEBUG] KMS Grant list truncated, getting next page via marker: %s", *out.NextMarker)
+		return findKmsGrantById(conn, keyId, grantId, out.NextMarker)
+	}
+
+	return nil, NewKmsGrantMissingError(fmt.Sprintf("[DEBUG] Grant %s not found for key id: %s", grantId, keyId))
+}
+
+// Can't have both constraint options set:
+// ValidationException: More than one constraint supplied
+// NB: set.List() returns an empty map if the constraint is not set, filter those out
+// using len(v) > 0
+func kmsGrantConstraintsIsValid(constraints *schema.Set) bool {
+	constraintCount := 0
+	for _, raw := range constraints.List() {
+		data := raw.(map[string]interface{})
+		if v, ok := data["encryption_context_equals"].(map[string]interface{}); ok {
+			if len(v) > 0 {
+				constraintCount += 1
+			}
+		}
+		if v, ok := data["encryption_context_subset"].(map[string]interface{}); ok {
+			if len(v) > 0 {
+				constraintCount += 1
+			}
+		}
+	}
+
+	if constraintCount > 1 {
+		return false
+	}
+	return true
+}
+
+func expandKmsGrantConstraints(configured *schema.Set) *kms.GrantConstraints {
+	if len(configured.List()) < 1 {
+		return nil
+	}
+
+	var constraint kms.GrantConstraints
+
+	for _, raw := range configured.List() {
+		data := raw.(map[string]interface{})
+		if contextEq, ok := data["encryption_context_equals"]; ok {
+			constraint.SetEncryptionContextEquals(stringMapToPointers(contextEq.(map[string]interface{})))
+		}
+		if contextSub, ok := data["encryption_context_subset"]; ok {
+			constraint.SetEncryptionContextSubset(stringMapToPointers(contextSub.(map[string]interface{})))
+		}
+	}
+
+	return &constraint
+}
+
+func concatStringMap(m map[string]*string, sep string) string {
+	var strList []string
+	for k, v := range m {
+		strList = append(strList, k, *v)
+	}
+	return strings.Join(strList, sep)
+}
+
+// The hash needs to encapsulate what type of constraint it is
+// as well as the keys and values of the constraint.
+func resourceKmsGrantConstraintsHash(v interface{}) int {
```

Comment:
I need to fix this up. As of now, I know I at least need to sort the maps so that the constraint gets consistent hashes. I couldn't find any common utilities for this.

Reply:
Fixed.